### PR TITLE
[Fiber] Ensure `srcset` and `src` are assigned last on `img` instances

### PR DESCRIPTION
**Pull request by gnoff**
Safari has a behavior (bug) where when you consturct an Image in javascript if you set srcset before properties for `sizes` the brwoser will download the largest image size because it starts to load before you communicate the sizes information.

https://x.com/OliverJAsh/status/1812408504444989588?t=CVHPqBaUiF5-6DBPGERTDA

There are likely other combinations or property order assignment that can cause problems such as setting crossorigin after assigning src or srcset. Conceptually we should withold the src and srcSet from the Image instance until last so all relevant other properties can be assigned before actually initiating any network activity.

This is an unforunate amount of code for what is realistically a bug in Browsers but it should allow us to avoid weird regressions depending on prop object order.

I didn't change the preload prop order because I don't believe preload links have the same issue (they are not fetched as eagerly I believe).

One nice benefit of this change though is the img case can move higher in the switch which is likely optimal given it's a relatively common tag. Previously it was as low as it was because it was part of the void element set so it couldn't be elevated without elevating less common tags

### CODE REVIEW

File: packages/react-dom-bindings/src/client/ReactDOMComponent.js [hunk 1]
```diff
@@ -1031,6 +1031,53 @@ export function setInitialProperties(
...
       break;
     }
+    // img tags previously were implemented as void elements with non delegated events however Safari (and possibly Firefox)
+    // begin fetching the image as soon as the `src` or `srcSet` property is set and if we set these before other properties
+    // that can modify the request (such as crossorigin) or the resource fetch (such as sizes) then the browser will load
+    // the wrong thing or load more than one thing. This implementation ensures src and srcSet are set on the instance last
+    case 'img': {
+      listenToNonDelegatedEvent('error', domElement);
+      listenToNonDelegatedEvent('load', domElement);
+      // Mostlye a port of Void Element logic with special casing to ensure srcset and src are set last
```
Comment (kassens) [inline packages/react-dom-bindings/src/client/ReactDOMComponent.js:1041]: nit
```suggestion
      // Mostly a port of Void Element logic with special casing to ensure srcset and src are set last
```

**Review by kassens — COMMENTED** [review]
While you're here, also work around the [second mentioned bug](https://x.com/kurtextrem/status/1812414413799911690) of `loading=lazy` needed before `src`?

File: packages/react-dom-bindings/src/client/ReactDOMComponent.js [hunk 2]
```diff
@@ -1031,6 +1031,53 @@ export function setInitialProperties(
       // Fast track the most common tag types
       break;
     }
+    // img tags previously were implemented as void elements with non delegated events however Safari (and possibly Firefox)
```
Comment (eps1lon) [inline packages/react-dom-bindings/src/client/ReactDOMComponent.js:1034]: Safari truly is the new IE.

**Review by eps1lon — APPROVED** [review]

**Comment by sophiebits** [comment]
Does this fix this bug too?
* #20682 
I guess the img tag needs to be appended to its parent before .src is set.

**Comment by sophiebits** [comment]
And does this need to do something with srcset:

https://github.com/facebook/react/blob/1f60a41801bde7fa6970c8cf36019248c2ee8996/packages/react-dom-bindings/src/client/ReactFiberConfigDOM.js#L696

**Comment by gnoff** [comment]
@sophiebits 

>And does this need to do something with srcset:

funny that's the first contribution I made to React I think :) I suspect yes it probably does need to include srcSet as well. I'll land in another PR

>I guess the img tag needs to be appended to its parent before .src is set.

Yeah this is something I thought about when adding this. The composition with a picture parent is probably a problem and needs more consideration. Unfortunately it's not as simple a change so I'm not going to do it now.

I wonder if we should take a stronger stance on `<picture>` and say that it can only have "simple" children and just do innerHTML or something. That way we can avoid all the property and parent assignment ordering without having to worry about each browsers idiosyncracies.

**Comment by gnoff** [comment]
@kassens 

>While you're here, also work around the [second mentioned bug](https://x.com/kurtextrem/status/1812414413799911690) of loading=lazy needed before src?

this should solve that issue too b/c it makes sure src and srcSet go last

**Comment by kassens** [comment]
> @kassens
> 
> > While you're here, also work around the [second mentioned bug](https://x.com/kurtextrem/status/1812414413799911690) of loading=lazy needed before src?
> 
> this should solve that issue too b/c it makes sure src and srcSet go last

D'oh, of course.

**Pull request merged**